kernel/Kconfig

```diff
@@ -872,6 +872,23 @@ config MP_NUM_CPUS
 	  Number of multiprocessing-capable cores available to the
 	  multicpu API and SMP features.
 
+config MP_TOTAL_NUM_CPUS
+	int "Total number of CPUs/cores"
+	default 1
+	range 1 5
+	help
+	  Number of cores in the system including those not used by Zephyr.
+
+config SMP_BASE_CPU
+	int "Starting CPU number"
+	default 0
+	help
+	  This option sets the starting cpu id for the SMP core.
+	  For RISC-V systems such as MPFS and FU540 this would be set to 1 to
+	  skip the E51 hart 0 as it is not usable in SMP configurations.
+	  This can also be used in conjunction with MP_NUM_CPUS to select a
+	  subset of available CPUs to use for Zephyr.
```

Comment:
I don't think these belong here.  The idea of the Zephyr CPU ID is that it's an abstraction.  Obviously every platform has its own rules about exactly what a "core" is and how they're indexed.  Some are simple, e.g. intel_adsp devices have a processor ID that matches the Zephyr one.  Others get weirder (esp32 puts odd numbers in its PRID registers).  Some are genuinely obtuse (addressing for x86 cores in a multi-socket board with hyperthreading and separate performance/efficiency cores).

But Zephyr doesn't care.  The Zephyr rules are:

* CPUs start from zero and go to N
* CPU zero is the one that enters z_cstart()
* CPUs 1+ don't do anything detectable until arch_start_cpu() is called for them.
* You can choose not to use a suffix of those cores for testing or efficiency reasons via MP_NUM_CPUS.

Everything else is left to the platform.  If you have a core that shouldn't appear to zephyr, don't put it in the list and don't give it an index.  If your useful cores are noncontiguous, use a lookup table or other data structure to convert.

But I don't see why Zephyr should ever need to know about CPUs in the system that it won't use.  Lots of data structures in the kernel are arrays of [CONFIG_MP_NUM_CPUS], and that indexing gets complicated if we need to track "fake" cores.

Reply:
Most of the areas in the system where the Kernel uses arrays indexed by CPU are not effected by this with the exception of the interrupt stacks and the issue I had there was that the startup code needs to be built in such a way that if it does get executed by a CPU that is not going to be running Zephyr, we need to park it safely which means we need a stack for it and the interrupt stacks are the default ones used at startup. The rest of this is hidden in the arch specific parts of the system.
No matter what scheme you use to limit the number of cores available for Zephyr, you need to be able to configure the Zephyr application appropriately so it has to rear its ugly head at some point. It has been suggested that the device tree might be where some of this should be handled but for a first cut I looked at kconfig because it provides a relatively simple and understandable way of implementing the configurability compared the DTS.

Reply:
The question is more "Why should intel_adsp or x86 platforms pay the cost of the added complexity of this CPU indexing scheme?" and "Do you really want to see this extended to work for multi-socket NUMA hardware and be forced to deal with that on PolarFire too?  Because that is where it is going if you put it in the core API."

Obviously yes, the app needs to configure it.  So let's put the cost on the apps that need the platform and not everyone.   And if we're going to pick an API that goes into the core kernel, let's be absolutely sure it's going to work for everyone and not just PolarFire.

I'm still unconvinced, sorry.